### ISSUE TITLE
Mistake In MovingPoint::getCoord?

### ISSUE THREAD

**Issue by Caravan0** [libspatialindex/libspatialindex#175]
Line 184 in libspatialindex/src/spatialindex/MovingPoint.cc

The range is discontinued.

Original function is
```
double MovingPoint::getCoord(uint32_t d, double t) const
{
	if (d >= m_dimension) throw Tools::IndexOutOfBoundsException(d);

	if (t >= m_endTime) return m_pCoords[d] + m_pVCoords[d] * (m_endTime - m_startTime);
	else if (t <= m_startTime) return m_pCoords[d] + m_pVCoords[d] * m_startTime;
	else return m_pCoords[d] + m_pVCoords[d] * (t - m_startTime);
}
```

Maybe this function should be 
```
double MovingPoint::getCoord(uint32_t d, double t) const
{
	if (d >= m_dimension) throw Tools::IndexOutOfBoundsException(d);

	if (t >= m_endTime) return m_pCoords[d] + m_pVCoords[d] * (m_endTime - m_startTime);
	else if (t <= m_startTime) return m_pCoords[d];
	else return m_pCoords[d] + m_pVCoords[d] * (t - m_startTime);
}
```

**Comment by Caravan0** [comment]
Another mistake in MovingRegion::containsRegionAtTime?

Line 655 in libspatialindex/src/spatialindex/MovingRegion.cc

the result of getExtrapolatedHigh is compared with itself.

Original function is
`
// does not work with degenerate time-intervals.
bool MovingRegion::containsRegionAtTime(double t, const MovingRegion& r) const
{
	if (m_dimension != r.m_dimension) throw Tools::IllegalArgumentException("containsRegionAtTime: MovingRegions have different number of dimensions.");

	// do they contain the time instant?
	if (! (m_startTime <= t && t < m_endTime && r.m_startTime <= t && t < r.m_endTime)) return false;

	for (uint32_t cDim = 0; cDim < m_dimension; ++cDim)
	{
		if (getExtrapolatedLow(cDim, t) > r.getExtrapolatedLow(cDim, t) || getExtrapolatedHigh(cDim, t) < getExtrapolatedHigh(cDim, t)) return false;
	}
	return true;
}

`

Maybe this function should be

`

bool MovingRegion::containsRegionAtTime(double t, const MovingRegion& r) const
{
	if (m_dimension != r.m_dimension) throw Tools::IllegalArgumentException("containsRegionAtTime: MovingRegions have different number of dimensions.");

	// do they contain the time instant?
	if (! (m_startTime <= t && t < m_endTime && r.m_startTime <= t && t < r.m_endTime)) return false;

	for (uint32_t cDim = 0; cDim < m_dimension; ++cDim)
	{
		if (getExtrapolatedLow(cDim, t) > r.getExtrapolatedLow(cDim, t) || getExtrapolatedHigh(cDim, t) < r.getExtrapolatedHigh(cDim, t)) return false;
	}
	return true;
}

`

**Comment by Caravan0** [comment]
I am almost a green hand in github and know little about 'insert code'.
Fortunately, what I want to emphasize is in the gray frame.

**Comment by Caravan0** [comment]
Third mistake in Index::SetIndexVariant?
Line 52 in libspatialindex/include/spatialindex/capi/Index.h and line 360 in libspatialindex/src/capi/Index.cc
the input parameter of SetIndexVariant(RTStorageType v) should be a RTIndexVariant instead of a RTStorageType.
Original function is
```
void SetIndexVariant(RTStorageType v);
void Index::SetIndexVariant(RTStorageType v)
{
	using namespace SpatialIndex;
	Tools::Variant var;

	if (GetIndexType() == RT_RTree) {
		var.m_val.ulVal = static_cast<RTree::RTreeVariant>(v);
		m_properties.setProperty("TreeVariant", var);
	} else if (GetIndexType() == RT_MVRTree) {
		var.m_val.ulVal = static_cast<MVRTree::MVRTreeVariant>(v);
		m_properties.setProperty("TreeVariant", var);
	} else if (GetIndexType() == RT_TPRTree) {
		var.m_val.ulVal = static_cast<TPRTree::TPRTreeVariant>(v);
		m_properties.setProperty("TreeVariant", var);
	}
}
```
Maybe this function should be
```
void SetIndexVariant(RTIndexVariant v);
void Index::SetIndexVariant(RTIndexVariant v)
{
	using namespace SpatialIndex;
	Tools::Variant var;

	if (GetIndexType() == RT_RTree) {
		var.m_val.ulVal = static_cast<RTree::RTreeVariant>(v);
		m_properties.setProperty("TreeVariant", var);
	} else if (GetIndexType() == RT_MVRTree) {
		var.m_val.ulVal = static_cast<MVRTree::MVRTreeVariant>(v);
		m_properties.setProperty("TreeVariant", var);
	} else if (GetIndexType() == RT_TPRTree) {
		var.m_val.ulVal = static_cast<TPRTree::TPRTreeVariant>(v);
		m_properties.setProperty("TreeVariant", var);
	}
}
```

**Comment by hobu** [comment]
Can you please make a PR that includes all of these fixes? I don't think the MVR variant has been exercised that much.

**Comment by hobu** [comment]
Fixed in #208
